### PR TITLE
Documentation overhaul - Fixing formatting issues, autodoc implementation, and added variables for versions.

### LINKED ISSUE #3639: Documentation updates/refactor
## Description

The [Hypha documentation](https://docs.hypha.app/) is largely out of date and at times disjointed. As adopters of the application continue to grow and the code base gets older + larger, updating the documentation, making it easily maintainable, and regularly reviewing it for accuracy is critical.

## To-Do

- [x] All mentioned software versions should be made into [variables](https://mkdocs-macros-plugin.readthedocs.io/en/latest/#getting-started) so references in the docs are automatically updated.
- [x] Code formatting should be more consistent, and easily allow the user to copy & paste (ie. `$ ls` -> `ls`).
- [x] Installation tutorials should be retested with the latest OS versions.
- [x] Re-evaluate that documentation formatting follows the [Diataxis](https://diataxis.fr/) principles
- [x] Establish a time frame that ensures docs are reviewed on a regular basis (ie. Quarterly review & testing)
- [x] #3080
- [x] Add descriptions for [apply site](https://docs.hypha.app/getting-started/architecture/#apply-site) & [wagtail admin](https://docs.hypha.app/getting-started/architecture/#wagtail-admin)
- [x] Review and document [configuration](https://docs.hypha.app/setup/administrators/configuration/) feature flags
- [x] Review screenshots for accuracy (ie. [Using Wagtail Admin](https://docs.hypha.app/user-guides/creating-a-user-account/#using-wagtail-admin))
- [x] Update for password-less login features (ie. [How to login](https://docs.hypha.app/user-guides/creating-a-user-account/#using-wagtail-admin))
- [x] Initial [docstring documentation](https://pypi.org/project/mkgendocs/) implementation
- [x] [Setup notifications](https://docs.hypha.app/setup/setup-notifications/) needs to be populated
- [ ] Document the public web API
- [x] Update vulnerability disclosure instructions

### PR DESCRIPTION
**Pull request by wes-otf**
## Description

Closes #3639. This is a bit of an overhaul of the docs to make them more easy to utilize. This changes include:
- Removal of overhanging public aspects of the documentation
- Reformatting of terminal commands to allow for easy copy & pasting
- Updating points of contact
- Screenshot Updates
- Automated code API documentation. [Google's docstring style](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings) should be used.
- Added variables for things like software versions & packages.
  - Unless python & node version are manually specified, when docs are built these will be pulled from [`.nvmrc`](https://github.com/HyphaApp/hypha/blob/main/.nvmrc) & [`.python-version`](https://github.com/HyphaApp/hypha/blob/main/.python-version)
- Consolidated the production and development guides into one location (both under `Setup` rather than one under `Getting Started` and one under `Setup`), and updated them to be similar to one another.
- Added different steps for getting Hypha development deployed under multiple OSs

## Screenshots

*API Documentation*
Screenshot 2024-01-23 at 10 48 58

## Test Steps

This would just require developer verification that docs are being built and generated as expected. Specifically automatic versioning and API documentation is being generated.

### CODE REVIEW

**Comment by wes-otf** [comment]
[0522725](https://github.com/HyphaApp/hypha/pull/3734/commits/052272541f13ce8ba9d31e7218a69091a02ebcd0]) was made on accident and was removed in a following commit

File: mkdocs.yml [hunk 1]
```diff
@@ -40,8 +40,33 @@ theme:
 plugins:
   - awesome-pages
   - search
+  - macros:
+      module_name: docs/scripts/doc_macros
+  - mkdocstrings:
```
Comment (theskumar) [inline mkdocs.yml:45]: > Automated code API documentation. [Google's docstring style](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings) should be used.

This has significantly increased the `mkdocs serve` time to refresh or just being useable? Would be possible to have this as optional while writing docs. Or provide an alternative to preview docs quickly?

File: mkdocs.yml [hunk 2]
```diff
@@ -40,8 +40,33 @@ theme:
...
+            merge_init_into_class: true
+            signature_crossrefs: true
+            show_symbol_type_heading: true
+            show_symbol_type_toc: true
   - git-revision-date-localized:
       type: timeago
+  - gen-files:
+      scripts:
+        - docs/scripts/gen_ref_pages.py
+  - section-index
```
Comment (theskumar) [inline mkdocs.yml]: This is causing the section get hidden in the navigation section itself, unless there is going to be `index.md` in each of the folder with some over about the section, I believe will cause bit of confusion to readers. 

image

File: requirements-docs.txt [hunk 3]
```diff
@@ -1,3 +1,7 @@
 mkdocs-material==9.4.6
```
Comment (theskumar) [inline requirements-docs.txt]: Upgrade to `9.5.5`? it seems to have some nices fixes compared to this version.

https://squidfunk.github.io/mkdocs-material/changelog/

File: requirements-docs.txt [hunk 4]
```diff
@@ -1,3 +1,7 @@
...
 mkdocs-awesome-pages-plugin==2.9.2
 mkdocs-git-revision-date-localized-plugin==1.2.1
+mkdocs-macros-plugin
+mkdocstrings[python]
+mkdocs-gen-files
+mkdocs-section-index
```
Comment (theskumar) [inline requirements-docs.txt]: I would suggest freezing the dependencies so we don't get accidentally broken docs.

File: mkdocs.yml (hunk 2 above)
Comment (theskumar) [reply mkdocs.yml]: Without `section-index` plugin. 

Screenshot 2024-01-25 at 6  19 21@2x

File: mkdocs.yml (hunk 1 above)
Comment (theskumar) [reply mkdocs.yml:45]: Looking deeper this seems to be easy doable if the `gen_ref_pages.py` uses a environment variable based flag to generate the initial reference files. Then enable that during build only by default. If someone want to enable during `mkdocs serve` it should be posssible too. 

```shell
GEN_REF_PAGES=True mkdocs build        # Generate reference docs, use in CI/CD
```

```shell
mkdocs serve        # skip reference docs, to be used locally / `Makefile` 
```

File: mkdocs.yml (hunk 2 above)
Comment (theskumar) [reply mkdocs.yml]: theme feature `navigation.indexes` (already enabled) probably serves the same purpose, but doesn't link the section header, if the `index.md` is missing. 

https://squidfunk.github.io/mkdocs-material/setup/setting-up-navigation/#section-index-pages

File: mkdocs.yml (hunk 1 above)
Comment (theskumar) [reply mkdocs.yml:45]: Alternatively, (better option) have `enabled` flag added to the `mkdocsstrings` config 

```yaml
  - mkdocstrings:
      enabled: !ENV [GEN_REF_PAGES, false]
      ...
```

File: docs/scripts/gen_ref_pages.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        continue
+
+    # Create the required files on build with mkdocs_gen_files, see:
+    # https://oprypin.github.io/mkdocs-gen-files/
+    if len(parts) > 0:
+        with mkdocs_gen_files.open(full_doc_path, "w") as fd:
+            identifier = f"::: hypha.{'.'.join(parts)}"
+            fd.write(identifier)
+
+    mkdocs_gen_files.set_edit_path(full_doc_path, path)
```
Comment (theskumar) [inline docs/scripts/gen_ref_pages.py]: Faster/concurrent version, also removes "migrations" folder from being processed.

```suggestion
import concurrent.futures
from pathlib import Path

import mkdocs_gen_files

# The code source directory
src = Path(__file__).parent.parent.parent / "hypha"
skip_folders = ["public", "tests", "migrations"]

def process_path(path):
    module_path = path.relative_to(src).with_suffix("")
    doc_path = path.relative_to(src).with_suffix(".md")
    full_doc_path = Path("references/API", doc_path)

    parts = tuple(module_path.parts)

    # Don't document any code in public/ or tests/ directories
    for skip_folder in skip_folders:
        if skip_folder in parts:
            return

    if parts[-1] == "__init__":
        parts = parts[:-1]
        doc_path = doc_path.with_name("index.md")
        full_doc_path = full_doc_path.with_name("index.md")
    elif parts[-1] in ["__main__"]:
        return

    # Create the required files on build with mkdocs_gen_files, see:
    # https://oprypin.github.io/mkdocs-gen_files/
    if len(parts) > 0:
        with mkdocs_gen_files.open(full_doc_path, "w") as fd:
            identifier = f"::: hypha.{'.'.join(parts)}"
            fd.write(identifier)

    mkdocs_gen_files.set_edit_path(full_doc_path, path)

def generate_pages():
    with concurrent.futures.ThreadPoolExecutor() as executor:
        list(executor.map(process_path, sorted(src.rglob("*.py"))))

generate_pages()
```

File: docs/setup/deployment/production/stand-alone.md [hunk 6]
```diff
@@ -62,14 +147,18 @@ To begin with, set the `export SECURE_SSL_REDIRECT=false` to prevent SSL redirec
...
 
-* `npm run build`
-* `python manage.py collectstatic --noinput`
-* `python manage.py createcachetable`
-* `python manage.py migrate --noinput`
-* `python manage.py clear_cache --cache=default --cache=wagtailcache`
-* `python manage.py createsuperuser`
-* `python manage.py wagtailsiteupdate server.domain apply.server.domain 80`
-* `python manage.py runserver` \(runs development server at [http://127.0.0.1:8000](http://127.0.0.1:8000)\)
+```console
```
Comment (theskumar) [inline docs/setup/deployment/production/stand-alone.md]: `console` stands for bash session. I believe the intension here is to have these as copy/paste shell script. Use `shell` marker instead?

Comment (theskumar) [reply docs/setup/deployment/production/stand-alone.md]: Screenshot 2024-01-25 at 7  57 26@2x

**Comment by wes-otf** [comment]
@theskumar This feedback is fantastic, thanks for taking the time to go over it all! I'll be handling/replying to all of these between today & tomorrow.

File: docs/scripts/gen_ref_pages.py (hunk 5 above)
Comment (wes-otf) [reply docs/scripts/gen_ref_pages.py]: This definitely makes it a lot snappier! Great suggestion.

File: mkdocs.yml (hunk 2 above)
Comment (wes-otf) [reply mkdocs.yml]: Yeah you're totally right, I didn't even notice `section-index` was duplicating the `navigation.indexes` functionality. I tossed `section-index` and all seems to work as expected from what I've seen!

File: mkdocs.yml (hunk 1 above)
Comment (wes-otf) [reply mkdocs.yml:45]: I also added an env check to gen_ref_files as that was still running w/o mkdocstrings being enabled.

File: .github/workflows/deploy-docs.yml [hunk 7]
```diff
@@ -24,6 +24,7 @@ jobs:
       - name: Deploy docs
         uses: mhausenblas/mkdocs-deploy-gh-pages@master
         env:
+          GEN_REF_PAGES: true
```
Comment (wes-otf) [inline .github/workflows/deploy-docs.yml:27]: @frjo @theskumar I believe this is the right place for this? Was between this & putting in one level up under the ubuntu-latest `env`.

Comment (theskumar) [reply .github/workflows/deploy-docs.yml:27]: I believe either would work. Feels right to me.

**Review by theskumar — APPROVED** [review]
💯

**Comment by frjo** [comment]
Excellent work here!

I was wondering around the change `apply.server.domain` to `server.domain` without a subdomain.

While Hypha no longer have a public site I think most implementors will install Hypha on a subdomain. Root domain (and www) most likely point to their main web site.

I therefore think it is most helpful to show all examples with a subdomain.

Ideas what the best example subdomain to use is? Is it `apply.` or `hypha.` or something else?

**Comment by wes-otf** [comment]
yeah that's a great point - probably in the config commands too then right? I think `apply.` is a good stand in. I'll revert now!

**Comment by wes-otf** [comment]
So I updated only the production side of things - to me it doesn't make much sense to have the subdomain be for local development too, but let me know your thoughts! I also added a note explaining why this is:
Screenshot 2024-02-06 at 16 20 10

**Review by frjo — APPROVED** [review]

**Comment by frjo** [comment]
@wes-otf  This is ready to merge in I think. Anything you want to add first?

**Comment by wes-otf** [comment]
@frjo I'm all set! I think it's ready to go.

**Pull request merged**